lib/Dialect/RTL/Ops.cpp

```diff
@@ -300,6 +300,23 @@ void ConstantOp::build(OpBuilder &builder, OperationState &result,
   build(builder, result, APInt(numBits, (uint64_t)value, /*isSigned=*/true));
 }
 
+/// Flattens `opInputs`, and inserts the flattened inputs to the Nth index of
+/// the original inputs. This is used when flattening in the canonicalization
+/// pass. Example: op(1, 2, op(3, 4), 5) -> op(1, 2, 3, 4, 5)
+template <typename Inputs>
+auto flattenNthInput(const Inputs &inputs, const Inputs &opInputs, size_t N) {
```

Comment:
I thought you'd pull the whole "for loop" out, and template the function on the Operation type.   I don't think this needs to be templated on `Inputs`, since that type should be the same for all of them.

Please mark this function static.

nit: if you keep it this way, please rename 'N' to something like 'splitNo' or something like that.  If you keep 'n', please lower case it.

Reply:
I thought about pulling out the entire loop, but then it would require (a) pulling in the rewriter and (b) determining whether it was a `success()` or a `failure()`. We could simply set the last pass to be flattening, i.e.
```
 LogicalResult matchAndRewrite(AndOp op,
                                  PatternRewriter &rewriter) const override {
   ...
   return flattenNthInput(inputs, rewriter); // returns success() or failure() depending on if there is an op to flatten.
}
```
However, I didn't want to solidify it as the last pass. I didn't know if there'd be any future issues of ordering, where maybe one pass should come before another for some reason or other.

Reply:
Fixed static and changed name of `N`. I also wasn't sure whether a `mlir::OperandRange` was a view of a container or the container itself. Looking at other uses, it seems it's passed by value, so I mimicked this.

For this and everything above, if you disagree I will be happy to change it.

Reply:
You don't think something like:

`if (tryFlatteningOperands(op, rewriter)) return success();`

would work?

Reply:
That should work, let me brew something up and I'll request your review.